### ISSUE TITLE
Asp.Net Web Forms Site does not create Roslyn directory under bin folder

### ISSUE THREAD

**Issue by reynojh** [aspnet/RoslynCodeDomProvider#98]
We have a large application and a number of smaller apps that were based on Web Site Projects that do not contain a project file, that we cannot get the new Roslyn Compiler enabled for it.   Its not feasible to retro those sites into Web Applications, so we are looking for a way to use Roslyn with the existing projects.    

We are using Visual Studio 2019, and when we enable c# 6+ support it adds the nuget packages, adds the reference for the DotNetCOmpilerPlatform.  We do not check in the Rosyln folder (way to many projects to have duplicate copies), and we are expecting the Rosyln folder to be restored under the bin folder when we compile, but that is not happening.

Are there plans to fix this or a work around that we can use?

**Comment by reynojh** [comment]
The options of using a Fully Qualified directory path are problematic as well.  Our build servers don't always have a consistent path.  Our Local Debugging could happen and be ran from any one of our 45 environments  Drive:\Project\Release Drive:\Project\Feature Drive:\Project\Dev.

**Comment by stefankip** [comment]
I run into this same issue as well using Web Application projects (Umbraco). Cloning the project from git, restoring nuget packages and building the project doesn't create the `roslyn` folder inside the `bin` folder.

**Comment by StephenMolloy** [comment]
I have left this issue open, because it is a legitimate concern for folks using the "Web Site" option in VS, and I do like to come back and think about it with a fresh mind every now and then. But this is unlikely to be fixed in any way. Since "Web Sites" are project-less, they do not work with MSBuild like all other project types, and we rely on MSBuild to handle placement of the correct roslyn binaries in the correct location. Visual Studio has been de-emphasizing "Web Site" for quite some time and they are not investing in scenarios like this anymore, and I haven't been able to think of a simple way to address this. Particularly given the way nuget is moving away from using powershell scripts and encouraging using .targets and .props files for doing things like this... which is exactly why "Web Sites" don't work - because they don't play nicely with .targets and .props files.

tldr - We understand this is an inconvenient issue for folks. But it is unlikely to ever get fixed. :/

**Comment by stefankip** [comment]
I see you guys talking about Web Site projects, but what about Web Application projects? It shouldn't be a problem for those? Because I do have this issue with web app projects.

**Comment by StephenMolloy** [comment]
@stefankip, "Web Application" projects should work. Have you tried updating to the just-released 3.6.0 package? Does that not work for you? Please feel free to open a separate issue if you're finding issues with 3.6.0 in a Web Application project.

**Comment by stefankip** [comment]
Hmm it's at 2.0.1 currently. Think that's the default version Umbraco uses (or some other package).
I'll use 3.6.0 from now on and guess the issue will be gone.

**Comment by StephenMolloy** [comment]
Well, 3.6.0 was just released yesterday. And it's the next 'release' version after 2.0.1.  I wouldn't expect any major apps to have updated yet. ;)  

Anyway, I would expect to not see this particular issue described above with Web Application projects in 2.0.1 either. But there have been quite a few changes here and there between the two versions. Before digging into what might be causing your issue, it would be good to know if it still exists with the latest and greatest.

**Comment by stefankip** [comment]
I'm currently even unable to reproduce it with v2.0.1 to be honest.
Did a clean clone of one of my repos, but the roslyn folder gets restored fine right now, so I guess this can be closed safely :-)

**Comment by rolivares** [comment]
problem with my MVC5 app on publishing app (not on local debugging), foler is not copied. I'm using last version 3.6

**Comment by dmce** [comment]
Appreciating this is older, its still an issue with the latest Umbraco which has a dependency Microsoft.CodeDom.Providers.DotNetCompilerPlatform (>= 2.0.1 && < 2.999999.0) so you have to use 2.0.1

A fresh nuget install doesnt add the roslyn folder. I think there are a couple of ways around, like cleaning and unloading project etc, but you can `Update-Package Microsoft.CodeDom.Providers.DotNetCompilerPlatform -r` which will add it properly.

**Comment by ondrejpialek** [comment]
I wonder if the issues could be caused by the fact that the imported props file does not exist yet during initital build, before any NuGet packages are downloaded. Notice the Exists condition - the project will load fine but will not include the props if they are not downloaded and present in the `packages` folder.

```
<Import Project="..\packages\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.2.0.1\build\net46\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.props" Condition="Exists('..\packages\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.2.0.1\build\net46\Microsoft.CodeDom.Providers.DotNetCompilerPlatform.props')" />
```

When I do the following sequence of steps I end up with a functional website:

- Clone the repo (= the packages folder does not exist yet)
- Build (= will restore NuGet packages but no roslyn in bin)
- Reopen the solution (= causes the project to reinitialize and load the props that have been previously restored)
- Rebuild (= to force all output be regenerated, ordinary build will report no changes and no folders will be created)

Not very fond of this approach, but seems to work.

**Comment by StephenMolloy** [comment]
I can only refer back to this comment here: https://github.com/aspnet/RoslynCodeDomProvider/issues/98#issuecomment-656335348

Web "Sites" are at an inherent disadvantage and are being increasingly phased out. Even our attempts to work around
the issues we have with them not being msbuild-compatible are being challenged by nuget moving closer to msbuild
and away from package scripting. This particular issue just isn't going to get better. :(

(There were a couple other comments that suggest there might be issues for non-web"site" projects as well. There have
been some. Hopefully those are addressed in the latest 3.11 preview.)
